File: src/components/flocking/quadtree.js
import {Point, Rectangle, Circle} from '../../utils/math';

class Quadtree {
  constructor(capacity, rectangle) {
    this.rect = rectangle;
    this.cap = capacity;
    this.points = [];
    this.divided = false;
  }

  subdivide() {
    let nw = new Rectangle(this.rect.x, this.rect.y, this.rect.w / 2, this.rect.h / 2);
    this.northwest = new Quadtree(this.cap, nw);
    let sw = new Rectangle(this.rect.x, this.rect.y + this.rect.h / 2, this.rect.w / 2, this.rect.h / 2);
    this.southwest = new Quadtree(this.cap, sw);
    let ne = new Rectangle(this.rect.x + this.rect.w / 2, this.rect.y, this.rect.w / 2, this.rect.h / 2);
    this.northeast = new Quadtree(this.cap, ne);
    let se = new Rectangle(this.rect.x + this.rect.w / 2, this.rect.y + this.rect.h / 2, this.rect.w / 2, this.rect.h / 2);
    this.southeast = new Quadtree(this.cap, se);


  }

  insert(boid) {

    if (!this.rect.contains(boid.position)) {
      return false;
    }

    if (this.points.length < this.cap) {
      this.points.push(boid);
      return true;
    } else {
      if (!this.divided) {
        this.subdivide();
        this.divided = true;
      }
      if (this.northwest.insert(boid)) {
        return true;
      } else if (this.northwest.insert(boid)) {
        return true;
      } else if (this.northeast.insert(boid)) {
        return true;
      } else if (this.southeast.insert(boid)) {
        return true;
      } else if (this.southwest.insert(boid)) {
        return true;
      }
    }
  }

  draw(ctx) {
    ctx.beginPath()
    ctx.rect(this.rect.x, this.rect.y, this.rect.w, this.rect.h);
    ctx.lineWidth = 2;
    ctx.strokeStyle = '#101010';
    ctx.stroke();
    ctx.closePath();
    if (this.divided) {
      this.northwest.draw(ctx);
      this.northeast.draw(ctx);
      this.southeast.draw(ctx);
      this.southwest.draw(ctx);
    }
  }

querySquare(x, y, s, arr) {
    let found = [];
    if(arr){
      found = arr;
    }

    if (!this.rect.intersectSquare(s, x, y)) {
      return;
    } else {
      for (let p of this.points) {
        if (circle.contains(p.position)) {
          found.push(p);
        }
      }

      if (this.divided) {
        this.northeast.querySquare(x, y, s, found);
        this.northwest.querySquare(x, y, s, found);
        this.southeast.querySquare(x, y, s, found);
        this.southwest.querySquare(x, y, s, found);
      }

      return found;
    }
  }
  query(circle, arr) {
    let found = [];
    if(arr){
      found = arr;
    }

    if (!this.rect.intersectCircle(circle)) {
      return;
    } else {
      for (let p of this.points) {
        if (circle.contains(p.position)) {
          found.push(p);
        }
      }

      if (this.divided) {
        this.northeast.query(circle, found);
        this.northwest.query(circle, found);
        this.southeast.query(circle, found);
        this.southwest.query(circle, found);
      }

      return found;
    }

  }


}

export {
  Quadtree,
  Rectangle,
  Point,
  Circle
}

File: webgl/src/quadtree.rs
use crate::rendering::Rectangle as drawableRect;

use crate::simulations::Boid;
use web_sys::{WebGlBuffer, WebGlProgram, WebGlRenderingContext as GL, WebGlUniformLocation};

#[derive(Debug, Clone)]
pub struct Rectangle {
    pub x: f32,
    pub y: f32,
    pub width: f32,
    pub height: f32,
}

impl Rectangle {
    pub fn contains(&self, point: (f32, f32)) -> bool {
        point.0 >= self.x
            && self.x + self.width > point.0
            && point.1 <= self.y
            && self.y - self.height < point.1
    }

    pub fn intersectCircle(&self, circle: (f32, f32, f32)) -> bool {
        //x,y,radius
        let x = (circle.0 - self.x).abs();
        let y = (circle.1 - self.y).abs();

        if x > self.width / 2.0 + circle.2 {
            return false;
        };
        if y > self.height / 2.0 + circle.2 {
            return false;
        };

        if x <= self.width / 2.0 {
            return true;
        };
        if y <= self.height / 2.0 {
            return true;
        };

        let corner_dist = (x - self.width / 2.0).powi(2) + (y - self.height / 2.0).powi(2);

        return corner_dist <= circle.2.powi(2);
    }
}
pub struct Quadtree {
    rectangle: Rectangle,
    capacity: i16,
    points: Option<Vec<(f32, f32, usize)>>,
    divided: bool,
    nw: Option<Box<Quadtree>>, //boxes?
    sw: Option<Box<Quadtree>>,
    ne: Option<Box<Quadtree>>,
    se: Option<Box<Quadtree>>,
}

impl Quadtree {
    pub fn new(capacity: i16, rectangle: Rectangle) -> Self {
        Self {
            rectangle,
            capacity,
            nw: None,
            sw: None,
            ne: None,
            se: None,
            divided: false,
            points: None::<Vec<(f32, f32, usize)>>,
        }
    }

    pub fn reset(&mut self) {
        self.divided = false;
        self.nw = None;
        self.sw = None;
        self.ne = None;
        self.se = None;
    }

    fn subdivide(&mut self) {
        self.nw = Some(Box::new(Quadtree::new(
            self.capacity,
            Rectangle {
                x: self.rectangle.x,
                y: self.rectangle.y,
                width: self.rectangle.width / 2.0,
                height: self.rectangle.height / 2.0,
            },
        )));

        self.sw = Some(Box::new(Quadtree::new(
            self.capacity,
            Rectangle {
                x: self.rectangle.x,
                y: self.rectangle.y - self.rectangle.height / 2.0,
                width: self.rectangle.width / 2.0,
                height: self.rectangle.height / 2.0,
            },
        )));

        self.ne = Some(Box::new(Quadtree::new(
            self.capacity,
            Rectangle {
                x: self.rectangle.x + self.rectangle.width / 2.0,
                y: self.rectangle.y,
                width: self.rectangle.width / 2.0,
                height: self.rectangle.height / 2.0,
            },
        )));

        self.se = Some(Box::new(Quadtree::new(
            self.capacity,
            Rectangle {
                x: self.rectangle.x + self.rectangle.width / 2.0,
                y: self.rectangle.y - self.rectangle.height / 2.0,
                width: self.rectangle.width / 2.0,
                height: self.rectangle.height / 2.0,
            },
        )));

        // offload points from this qt to the children and turn points to none
        for point in self.points.as_ref().unwrap() {
            self.ne.as_mut().unwrap().insert(point.0, point.1, point.2);
            self.se.as_mut().unwrap().insert(point.0, point.1, point.2);
            self.nw.as_mut().unwrap().insert(point.0, point.1, point.2);
            self.sw.as_mut().unwrap().insert(point.0, point.1, point.2);
        }

        self.points = None;

        self.divided = true;
    }

    pub fn insert(&mut self, x: f32, y: f32, boid_index: usize) -> bool {
        if !self.rectangle.contains((x, y)) {
            false
        } else if self.divided {
            if self.nw.as_mut().unwrap().insert(x, y, boid_index) {
                true
            } else if self.ne.as_mut().unwrap().insert(x, y, boid_index) {
                true
            } else if self.sw.as_mut().unwrap().insert(x, y, boid_index) {
                true
            } else if self.se.as_mut().unwrap().insert(x, y, boid_index) {
                true
            } else {
                false
            }
        } else if self.points.is_none() {
            //if self.points.as_ref().unwrap().len() < self.capacity as usize {
            self.points = Some(Vec::new());
            let _ = self.points.as_mut().unwrap().push((x, y, boid_index));

            true
        } else if self.points.as_ref().unwrap().len() < self.capacity as usize {
            let _ = self.points.as_mut().unwrap().push((x, y, boid_index));

            true
        } else {
            if !self.divided {
                self.subdivide();
            }
            if self.nw.as_mut().unwrap().insert(x, y, boid_index) {
                true
            } else if self.ne.as_mut().unwrap().insert(x, y, boid_index) {
                true
            } else if self.sw.as_mut().unwrap().insert(x, y, boid_index) {
                true
            } else if self.se.as_mut().unwrap().insert(x, y, boid_index) {
                true
            } else {
                false
            }
        }
    }

    pub fn query(&self, circle: (f32, f32, f32)) -> Vec<usize> {
        let mut found: Vec<usize> = Vec::new();
        if self.divided {
            found.extend(self.ne.as_ref().unwrap().query(circle));
            found.extend(self.nw.as_ref().unwrap().query(circle));
            found.extend(self.se.as_ref().unwrap().query(circle));
            found.extend(self.sw.as_ref().unwrap().query(circle));
        }

        if self.points.is_none() || !self.rectangle.intersectCircle(circle) {
            return found;
        } else {
            for point in self.points.as_ref().unwrap() {
                if ((point.0 - circle.0).powi(2) + (point.1 - circle.1).powi(2)).sqrt() <= circle.2
                {
                    found.push(point.2);
                }
            }
        }
        return found;
    }

    pub fn renderroot(&self, gl: &GL, line: &drawableRect) {
        let color = [0.5, 0.5, 0.5, 1.0];

        line.render(
            &gl,
            self.rectangle.x,
            self.rectangle.y,
            self.rectangle.width,
            -0.007,
            color,
        );

        line.render(
            &gl,
            self.rectangle.x,
            self.rectangle.y - self.rectangle.width,
            0.004,
            self.rectangle.height,
            color,
        );

        line.render(
            &gl,
            self.rectangle.x,
            self.rectangle.y - self.rectangle.width,
            self.rectangle.width,
            0.007,
            color,
        );

        line.render(
            &gl,
            self.rectangle.x + self.rectangle.width,
            self.rectangle.y - self.rectangle.height,
            -0.004,
            self.rectangle.height,
            color,
        );

        if self.divided {
            self.nw
                .as_ref()
                .unwrap()
                .renderchild(&gl, (true, true), &line);
            self.ne
                .as_ref()
                .unwrap()
                .renderchild(&gl, (true, false), &line);
            self.se
                .as_ref()
                .unwrap()
                .renderchild(&gl, (false, false), &line);
            self.sw
                .as_ref()
                .unwrap()
                .renderchild(&gl, (false, true), &line);
        }
    }

    pub fn renderchild(&self, gl: &GL, dir: (bool, bool), line: &drawableRect) {
        let color = [0.5, 0.5, 0.5, 1.0];

        if dir.0 && dir.1 {
            //nw
            line.render(
                &gl,
                self.rectangle.x,
                self.rectangle.y - self.rectangle.height,
                self.rectangle.width,
                0.004,
                color,
            );
            line.render(
                &gl,
                self.rectangle.x + self.rectangle.width + 0.002, //0.002 is half the width of a line
                self.rectangle.y - self.rectangle.height,
                0.004,
                self.rectangle.height,
                color,
            )
        } else if dir.0 && !dir.1 {
            //ne
            line.render(
                &gl,
                self.rectangle.x,
                self.rectangle.y - self.rectangle.height,
                self.rectangle.width,
                0.004,
                color,
            )
        } else if !dir.0 && dir.1 {
            //se
            line.render(
                &gl,
                self.rectangle.x + self.rectangle.width + 0.002,
                self.rectangle.y - self.rectangle.height,
                0.004,
                self.rectangle.height,
                color,
            )
        }

        if self.divided {
            self.nw
                .as_ref()
                .unwrap()
                .renderchild(&gl, (true, true), &line);
            self.ne
                .as_ref()
                .unwrap()
                .renderchild(&gl, (true, false), &line);
            self.se
                .as_ref()
                .unwrap()
                .renderchild(&gl, (false, false), &line);
            self.sw
                .as_ref()
                .unwrap()
                .renderchild(&gl, (false, true), &line);
        }
    }
}
